controller/contour/controller.go

```diff
@@ -0,0 +1,123 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package contour
+
+import (
+	"context"
+	"fmt"
+	"k8s.io/apimachinery/pkg/types"
+
+	retryable "github.com/projectcontour/contour-operator/util/retryableerror"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/projectcontour/contour-operator/util/slice"
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	operatorv1alpha1 "github.com/projectcontour/contour-operator/api/v1alpha1"
+)
+
+const (
+	contourFinalizer string = "contour.operator.projectcontour.io/finalizer"
+
+	defaultContourNamespace = "projectcontour"
+)
+
+// Config holds all the things necessary for the controller to run.
+type Config struct {
+	// Image is the name of the Contour container image.
+	Image string
+}
+
+// Reconciler reconciles a Contour object.
+type Reconciler struct {
+	Config
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=operator.projectcontour.io,resources=contours,verbs=get;list;watch;create;update;patch;delete
```

Comment:
Do we need `create;update;patch;delete` access on the main resource?

Reply:
I'll change to `get;list;watch;update`. We'll most likely have future use cases that will require the additional verbs, but the they can be updated as needed.